### PR TITLE
Cache order year_months in options for performance.

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Caches the slow query which calculates year-months to populate the dates filter on order admin screen:

Screenshot 2024-07-29 at 3 24 50 PM

We cache the results in wp options table, since these results won't be invalidated soon. We validated the cache again every day since new day might also be a new month. So we update the cache to add this month if needed. Additionally, there might be bulk orders imported for previous days, so in this case, the cache will be invalid only till day is changed.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Make sure HPOS is authoritative.
1. Start with a new site without any orders. Go to the order admin page and verify that current year month is still displayed.
3. In a shop with existing orders (if you only have a new shop, add some orders and manually change the date to previous months), switch to this branch, go to the order admin page, refresh it, and check that the orders are populated as expected. 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]: I wonder if we could rearrange the order of things a bit to make these two extra database operations unnecessary (since the same rows get updated further down)...

**Review by coreymckrill — APPROVED** [review]
👍 Looks like a good optimization. In addition to the testing instructions, I tried changing an order date to ~3 years ago, and then invalidating the cache. Approving, but one comment below to potentially discuss before merging.

**Comment by vedanshujain** [comment]
@coreymckrill Modified the logic a bit to improve db calls + unit tests, can you take another look?

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		if ( false !== $cached_value ) {
```

It looks like if the query further down returns an empty array, that will still get stored in the cache, but then will be interpreted as false here, thus causing the query to run again the next time.

File: plugins/woocommerce/tests/php/src/Internal/Admin/Orders/ListTableTest.php
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/src/Internal/Admin/Orders/ListTableTest.php:1]: Would it be worth adding another test or two here to check that the current year/month always gets added?

Comment (vedanshujain) [reply plugins/woocommerce/tests/php/src/Internal/Admin/Orders/ListTableTest.php:1]: Added in https://github.com/woocommerce/woocommerce/pull/50066/commits/132a2fd5f4cf89ed847178b9fbb1d667e1331812

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]: Make sense, added here -> https://github.com/woocommerce/woocommerce/pull/50066/commits/e371fe04b67e541fc697bd9cf412886ead8e4513

**Comment by vedanshujain** [comment]
Since changes are to unit tests only, I will merge after CI passes (but let me know if there are more thoughts and would be happy to raise more PRs).

**Pull request merged**
